### PR TITLE
VimPlugins: update redirected repos

### PR DESCRIPTION
**Pull request by ryneeverett**
###### Motivation for this change
Many of the plugins in vim-plugin-names are out of date and redirect to their new github repos. This adds a flag to automatically update these, defines a process for committing those updates into nixpkgs, and performs the update.
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by jonringer — APPROVED** [review]
diff LGTM
commits LGTM

going to give some time for others to commit, ping me in 2-3 days if nothing happends

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: This seems surprising to me. Why do we want the first two parts of the URL? I would assume we need the last two parts instead:

```
github.com/owner/repo
```

-> First two would be `github.com` as the owner and `owner` as the repo.

Am I missing something?

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: The `"\n"` should probably be separate since it is always appended. For example:

```python
alias_spec = f" as {self.alias}\n" if self.alias else ""
plugin_line = "{owner}/{name}" + alias_spec + "\n"
```

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: This seems very brittle to me, can't we use the original line instead? This breaks on any whitespace change.

Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Or alternatively we could just rewrite the whole file and not bother replacing anything.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: This step would also be unnecessary if we'd just completely rewrite the file.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: It would be nicer to just generate the nice expressions in the first place. I suspect this step would be easy to miss and even easier to miss in review.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: If we suggest to do this anyway, why not default to `true`?

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [inline pkgs/misc/vim-plugins/update.py]: It would be amazing if we could automate this as well, though that's not a requirement for merge in my view (but a very nice to have).

**Review by timokau — CHANGES_REQUESTED** [review]
Great idea! I've left some feedback in the review. Also, did you check this with `mypy`?

Instead of adding aliases I would kind of prefer to just `throw` instead, but that would be a policy change and is definitely out of scope of this PR.

**Comment by ryneeverett** [comment]
I did indeed check with mypy.

> Instead of adding aliases I would kind of prefer to just throw instead, but that would be a policy change and is definitely out of scope of this PR.

Definitely! The aliases largely defeat the my motive to fix the redirections in the first place. One of the most troubling aspects of the vim plugin ecosystem is that there's very little attention when the owner/org changes.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse

github.com is the `netloc` in this scenario and we're only looking at the `path`.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: In your example if `self.alias` is truthy the resulting `plugin_line` would result in an extra newline: "{owner}/{name}\n\n".

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I'm not sure what you mean by a "whitespace change". The `vim-plugin-names` file is rewritten in the same run of `update.py` as the redirections are detected here so why would the file change?

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I'd rather invest effort into eliminating aliases.nix if possible.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I agree it would be good if `update.py` automatically sorted `vim-plugin-names`, though that might be out of scope for this PR.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Oh yes, I didn't mean to include the `\n` in the `alias_spec`. Also now that you mention it, I would rather not rely on things being "truthy" but instead check explicitly (`is None`).

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: I mean that this would break as soon as we'd have a line
```
owner/repo as  alias
```

(note the double space). The simplified solution would be to just remember the new lines (regardless of whether there was a redirect or not) and completely re-create the file instead of updating it. That would also simplify the sorting.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: It would be nice if multiple invocations were not necessary but I think this script would then need to automatically commit the update before updating the redirects. It seems like if this was all done at once in one commit it would be harder hard to review than if the updates were separated from the redirect updates as I suggest.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: I would propose to keep that file (or an equivalent file) but just add entries in the form of

```
old_name = throw "${old_name} was renamed to ${new_name}. Please use that instead" # added on yyyy-dd-mm
```

Since just doing the rename without any notification would be confusing to end users. We could then drop the throw lines once they're "old enough" (maybe 2 releases or something, see https://github.com/NixOS/rfcs/pull/33).

So either way we'd still need a file that keeps track of name changes.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I don't know exactly what the use case for the `--input-file` argument of this script is so I was cautious. It seems designed so one could maintain their own plugin list and it would be unexpected for this script to magically rewrite their file.

**Comment by timokau** [comment]
> Definitely! The aliases largely defeat the my motive to fix the redirections in the first place. One of the most troubling aspects of the vim plugin ecosystem is that there's very little attention when the owner/org changes.

Feel free to do create a follow-up PR/issue for that once this is done. Though owner/org changes would still remain unnoticed since the names are not namespaced (something which I also kind of wish wasn't the case).

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: `alias` is actually an empty string in this case, the truthiness of which is relied on in a couple other `alias or` clauses in this script. Do you prefer `if self.alias != ""`?

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I pushed a commit to normalize whitespace to deal with this particular issue. I recognize that this is still a bit flimsy but I would think it acceptable for a helper script, the results of which will be manually reviewed.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: I added a commit to throw errors like this rather than silently aliasing.

File: pkgs/misc/vim-plugins/update.py
Comment (Mic92) [reply pkgs/misc/vim-plugins/update.py]: I agree. We don't need an option here. Just make it the default.
If someone complains about the new behavior with a good reason, we can still add it back.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Yes, I definitely prefer explicit comparisons. "Truthiness" seems to directly contradict the python ethos of explicitness. It reduces readability, especially for people not familiar with python ("truthiness" differs between languages).

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: > I pushed a commit to normalize whitespace to deal with this particular issue.

Why didn't you go for the complete rewrite? Seems more robust, easier and you get a sorted result "for free".

> I recognize that this is still a bit flimsy but I would think it acceptable for a helper script, the results of which will be manually reviewed.

I'd be careful with the "manual review" statement. These big automated PRs tend to be skimmed for malicious intent, but details tend to be skipped. At least that's how I do it. They are simply too big and too repetitive for detailed review.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Wouldn't that actually reduce complexity?

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Yes, committing both changes at the same time is not ideal. But I think it is better than accidentally having the two files go out of sync, which I think is likely with the current model.

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Your point is valid, but this is a nixpkgs-internal tool and I don't think we need to prioritize backwards compatibility. If someone else is using this outside of nixpkgs, its their responsibility to keep up with upstream changes. If we wanted to advertise this outside of nixpkgs we'd make it into its own package (with semver).

File: pkgs/misc/vim-plugins/update.py
Comment (jonringer) [reply pkgs/misc/vim-plugins/update.py]: I agree that it would be nice, but I think it's out of scope of the PR, and can be done later.

Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: Now that the `--update-rewrite` flag is gone and we're modifying the input file without user input we might as well sort it while we're at it. As @timokau points out it is basically free.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: The file is rewritten every time now but I'm not sure if it's what you had in mind. It's not any more robust with respect to parsing the existing file.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: The flag is removed.

**Comment by ryneeverett** [comment]
I believe the last remaining issue is simplifying the update process i proscribe, how to automate it, and whether this is in scope or better done in another PR. Here's what is printed currently:

```py
        print(
            f"""\
Redirects have been detected and {input_file} has been updated. Please take the
following steps:
    1. Go ahead and commit just the updated expressions as you intended to do:
            git add {output_file}
            git commit -m "vimPlugins: Update"
    2. If any of the plugin names were changed, throw an error in aliases.nix:
            <oldName> = deprecateName "<oldName>" "<newName>"; # backwards compat, added YYYY-MM-DD
    3. Run this script again so these changes will be reflected in the
    generated expressions:
            ./update.py
    4. Commit {input_file} along with aliases and generated expressions:
            git add {output_file} {input_file} aliases.nix
            git commit -m "vimPlugins: Update redirects"
        """
        )
```

Thoughts:

1. Automatically modifying aliases.nix seems a bit fragile and messy. What if there was another generated file, say `deprecated-names.nix`, that aliases.nix read and concatenated with it's own aliases?
2. Should update.py automatically perform a `vimPlugins: Update` commit if it detects no modification in `vim-plugin-names`? This would simplify this process and the process of adding new plugins for contributors. We'd probably want the script to check if the git status is clean and exit immediately with a warning if it's not. Then the question is whether to hack it with subprocess or add a git python library to the shebang environment. (I would definitely think this is out of scope.)
3. Should I give up on separating plugin updates from redirection updates? Without this we could reduce the process to one `update.py` invocation without automating git commits.

File: pkgs/misc/vim-plugins/update.py
Comment (ryneeverett) [reply pkgs/misc/vim-plugins/update.py]: Done

**Comment by timokau** [comment]
> Thoughts:
> 
>     1. Automatically modifying aliases.nix seems a bit fragile and messy. What if there was another generated file, say `deprecated-names.nix`, that aliases.nix read and concatenated with it's own aliases?

That sounds like an excellent idea. Instead of a nix file, it would be
even nicer/easier to use a simpler format such as json here. We'd need
something that nix can parse, I'm not sure if something like that exists
though.

>     2. Should update.py automatically perform a `vimPlugins: Update` commit if it detects no modification in `vim-plugin-names`? This would simplify this process and the process of adding new plugins for contributors. We'd probably want the script to check if the git status is clean and exit immediately with a warning if it's not. Then the question is whether to hack it with subprocess or add a git python library to the shebang environment. (I would definitely think this is out of scope.)

I think that would be fine. I implemented that in the spotify updater
too. We should probably get at least a third opinion here though
(@Mic92, @jonringer).

>     3. Should I give up on separating plugin updates from redirection updates? Without this we could reduce the process to one `update.py` invocation without automating git commits.

That would be slightly less ideal, but also okay in my opinion. One
problem here is that someone could try to sneak in a malicious update,
but that's already the case anyway. At least it wouldn't survive the
next update.

Another (probably more complex) possibility would be to keep printing
the warning but only rewrite the `plugin-names` file if no plugins were
updated. My main issue with the current state is that the tree is left
in an inconsistent state and its easy to commit that by mistake.

**Comment by ryneeverett** [comment]
@timokau Can we hold off on deprecation automation and git automation for a separate PR? If so, can you approve this PR?

**Comment by jonringer** [comment]
Personal feelings

> Should update.py automatically perform a `vimPlugins: Update` commit if it detects no modification in `vim-plugin-names`? This would simplify this process and the process of adding new plugins for contributors. We'd probably want the script to check if the git status is clean and exit immediately with a warning if it's not. Then the question is whether to hack it with subprocess or add a git python library to the shebang environment. (I would definitely think this is out of scope.)

If we had it as part of some CI, then I think it would be important to create a git commit, however, I think it's less surprising to an individual if the changes are put into unstaged instead of a commit. If we did want a git commit, then I think it should be opted-in, such as `update.py --commit`

> Should I give up on separating plugin updates from redirection updates? Without this we could reduce the process to one `update.py` invocation without automating git commits.

We could do this as another PR

> @timokau Can we hold off on deprecation automation and git automation for a separate PR?

Agreed


I think it's a big improvement to have the updated repos, I'm fine with merging just those changes, and tackleing the automation/git history in another PR

**Comment by ryneeverett** [comment]
> If we had it as part of some CI, then I think it would be important to create a git commit, however, I think it's less surprising to an individual if the changes are put into unstaged instead of a commit.

The advantages of git integration are (1) separate commits for updates and redirections and (2) automating standardized commit messages. I don't see the point without committing.

**Comment by jonringer** [comment]
> The advantages of git integration are (1) separate commits for updates and redirections and (2) automating standardized commit messages. I don't see the point without committing.

I think I was coming from the perspective of these would be opt-in options (E.g `update.py --update-redirects) and not trying to both an update and redirect update at the same time. 

But yes, if the plugins and redirects were being updated, it would make sense to differentiate them in commits.

If it is the case that the git commits happen as part of the script, then please also update docs at https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/vim.section.md#adding-new-plugins-to-nixpkgs to reflect the new git behavior

**Comment by timokau** [comment]
> @timokau Can we hold off on deprecation automation and git automation for a separate PR? If so, can you approve this PR?

Do you commit to making those changes in a separate PR? Otherwise I fear that it'll never get done and we end up with an update process that is slightly worse than it is now (since it requires manual intervention). The result is better, but the process is more cumbersome. I'd like to avoid that at least long-term.

**Comment by ryneeverett** [comment]
> Do you commit to making those changes in a separate PR? Otherwise I fear that it'll never get done and we end up with an update process that is slightly worse than it is now (since it requires manual intervention). The result is better, but the process is more cumbersome.

I was planning on working on it but I wouldn't say I'm committed.

I'm not sure how this update process is any worse or more cumbersome, the manual intervention it suggests is only to do something that wasn't done at all before. If you don't do the manual intervention you'll either not commit the changes to `vim-plugin-names` and it will be the same as before, or you will commit them and they'll be updated eventually though it might escape notice.

**Comment by timokau** [comment]
> I'm not sure how this update process is any worse or more cumbersome -- the manual intervention it suggests is only to do something that wasn't done at all before. If you don't do the manual intervention you'll either not commit the changes to `vim-plugin-names` and it will be the same as before, or you will commit them and they'll be updated eventually though it might escape notice.

If you don't do any manual update you might

- End up with an inconsistent tree (though that's not too bad).
- Forget to add aliases / throw declarations. This breaks users configs in a confusing way.

And the big issue here is that it is very easy to miss that during update, and even easier to miss during review. I'd really like to avoid case (2) in particular.

Apparently nix does read json and TOML (grep the tree for `builtins.from` for examples). So it would be possible to dump a simple alias / deprecation notice dict into a json file and then read it from nix.

**Comment by ryneeverett** [comment]
@timokau Fair enough, though your pessimism that the review process wouldn't catch package renames is concerning.

**Comment by timokau** [comment]
Well, imagine you had to review https://github.com/NixOS/nixpkgs/pull/79911. How confident are you that you would catch a rename or an invalid update? Now imagine you're doing that weekly.

Really the only reliable way to make sure these PRs are really okay is to fully automate the updates and ideally have them run by a trusted entity.

**Comment by ryneeverett** [comment]
Agreed but I'd blame the automated part of the review too on a broader nixpkgs level too. Seems like ofborg should be adding a `package-deleted` tag or something when a package name disappears from evaluation.

**Comment by timokau** [comment]
That might be useful, but somebody would need to implement that ;) Its also less useful in most cases, since most PRs touch just one package and in that case a deletion / rename would be more obvious.

**Comment by jonringer** [comment]
> Agreed but I'd blame the automated part of the review too on a broader nixpkgs level too. Seems like ofborg should be adding a `package-deleted` tag or something when a package name disappears from evaluation.

It adds the "clean up" label

**Comment by ryneeverett** [comment]
I pushed a commit automating deprecation. I'll have to look at it tomorrow because I managed to noob up the 3 lines of nix I wrote and it doesn't evaluate.

**Comment by ryneeverett** [comment]
It's working fine as long as I don't use any of the deprecated names, but if I add one of the deprecated names to my plugins I get:

> error: attribute 'rtp' missing, at /nix/store/k683jb89wgfr09kwfq16anirn5r95ln3-nixos-20.03beta695.b2935fbecea/nixos/pkgs/misc/vim-plugins/vim-utils.nix:222:59

I must have made a dumb mistake in `aliases.nix`. Any ideas?

**Comment by timokau** [comment]
This actually works perfectly for me with `gist-vim`

```
$ nix-build -E 'with import ./default.nix {}; neovim.override { configure = { packages.myVimPackage = with vimPlugins; { start = [ gist-vim ]; }; }; }'

error: gist-vim was renamed to vim-gist. Please update to vim-gist.
(use '--show-trace' to show detailed location information)
```

Which plugin where you trying to add? The error occurs because the plugin mechanism expects vim plugins to have an `rtp` attribute. That would happen if you add some non-plugin thing to your vim plugins. But since the deprecations `throw` (which immediately terminates) it shouldn't come so far.

**Comment by timokau** [comment]
It would be nice to store the date of the deprecation. That way, when and if we decide on a strategy, we can prune them after some time. This should work:

```patch
diff --git a/pkgs/misc/vim-plugins/aliases.nix b/pkgs/misc/vim-plugins/aliases.nix
index ac6ffc4c14d..194433fed7f 100644
--- a/pkgs/misc/vim-plugins/aliases.nix
+++ b/pkgs/misc/vim-plugins/aliases.nix
@@ -31,8 +31,8 @@ let
                               (checkInPkgs n alias)))
                      aliases;
 
-  deprecations = lib.mapAttrs (old: new: {
-    old = throw "${old} was renamed to ${new}. Please update to ${new}.";
+  deprecations = lib.mapAttrs (old: info: {
+    old = throw "${old} was renamed to ${info.new} on ${info.date}. Please update to ${info.new}.";
   }) (builtins.fromJSON (builtins.readFile ./deprecated.json));
 
 in
diff --git a/pkgs/misc/vim-plugins/deprecated.json b/pkgs/misc/vim-plugins/deprecated.json
index bb4ad18c3ec..197d6806fd7 100644
--- a/pkgs/misc/vim-plugins/deprecated.json
+++ b/pkgs/misc/vim-plugins/deprecated.json
@@ -1,6 +1,18 @@
 {
-    "gist-vim": "vim-gist",
-    "vim-jade": "vim-pug",
-    "vundle": "Vundle.vim",
-    "youcompleteme": "YouCompleteMe"
-}
\ No newline at end of file
+    "gist-vim": {
+        "new": "vim-gist",
+        "date": "2020-03-26"
+    },
+    "vim-jade": {
+        "new": "vim-pug",
+        "date": "2020-03-26"
+    },
+    "vundle": {
+        "new": "Vundle.vim",
+        "date": "2020-03-26"
+    },
+    "youcompleteme": {
+        "new": "YouCompleteMe",
+        "date": "2020-03-26"
+    }
+}
diff --git a/pkgs/misc/vim-plugins/update.py b/pkgs/misc/vim-plugins/update.py
index d513669d0f3..bbeef0889f4 100755
--- a/pkgs/misc/vim-plugins/update.py
+++ b/pkgs/misc/vim-plugins/update.py
@@ -420,13 +420,17 @@ def rewrite_input(input_file: Path, output_file: Path, redirects: dict):
     if redirects:
         lines = [redirects.get(line, line) for line in lines]
 
+        cur_date_iso = datetime.now().strftime("%Y-%m-%d")
         with open(DEPRECATED, "r") as f:
             deprecations = json.load(f)
         for old, new in redirects.items():
             old_name = old.split("/")[1].split(" ")[0].strip("\n")
             new_name = new.split("/")[1].split(" ")[0].strip("\n")
             if old_name != new_name:
-                deprecations[old_name] = new_name
+                deprecations[old_name] = {
+                    "new": new_name,
+                    "date": cur_date_iso,
+                }
         with open(DEPRECATED, "w") as f:
             json.dump(deprecations, f, indent=4, sort_keys=True)
 
```

**Comment by ryneeverett** [comment]
On my latest commit 1cb7fb91166aeb657cbc6a54e8f9473f30738d14 I get the rtp error for gist-vim though the previous commit b886ad90e7395d98d9c4956eef817e29bfce71a5 does yield the correct deprecation error. I agree on your analysis -- I don't see how that change is causing it to be parsed as a plugin when it should be a literal `throw` statement.

I can't reproduce your example:

```sh
$ nix-build -E 'with import ./default.nix {}; neovim.override { configure = { packages.myVimPackage = with vimPlugins; { start = [ gist-vim ]; }; }; }'
error: anonymous function at /home/ryne/projects/nixpkgs/pkgs/misc/vim-plugins/default.nix:2:1 called without required argument 'callPackage', at (string):1:6
```

I'd considered your approach of adding the date but couldn't think of any reason to bother pruning deprecations unless/until conflicts with new packages using the deprecated names arise. However I don't object and appreciate the additional structure imposed on `deprecated.json`. I'll merge your patch once I'm confident that the branch is working without it. Alternatively, feel free to push it into my branch yourself at any time.

**Comment by timokau** [comment]
I don't have any issue at https://github.com/NixOS/nixpkgs/commit/1cb7fb91166aeb657cbc6a54e8f9473f30738d14.

You need to execute that command from the root of your nixpkgs clone (or adapt the `import` statement to point to the `default.nix` at the root).

`nix-build -E 'with import ./default.nix {}; attrname` is essentially equivalent to `nix-build -A attrname`, except that it allows you to use arbitrary expressions instead of `attrname`.

**Comment by ryneeverett** [comment]
Thanks. I can reproduce and indeed have no problem with either commit using vim packages as you suggest. However, the latest commit does break the vim-plug implementation:

```sh
$ git checkout b886ad9 &>/dev/null && nix-build -E 'with import ./default.nix {}; neovim.override { configure = { plug.plugins = with vimPlugins; [ gist-vim ]; }; }'
error: vim-gist was renamed to gist-vim. Please update to gist-vim.
(use '--show-trace' to show detailed location information)
$ git checkout 1cb7fb9 &>/dev/null && nix-build -E 'with import ./default.nix {}; neovim.override { configure = { plug.plugins = with vimPlugins; [ gist-vim ]; }; }'
error: attribute 'rtp' missing, at /home/ryne/projects/nixpkgs/pkgs/misc/vim-plugins/vim-utils.nix:222:59
(use '--show-trace' to show detailed location information)
```

**Comment by timokau** [comment]
Ah, I've found the issue(s).

- `old = throw ...` will actually assign the `throw` to the literal `old`, i.e. `old` is not an expression. To get the behaviour you intended there it would have to be `"${old}" = throw ...`.
- `lib.mapAttrs` creates a new attribute set keyed by the old keys by default, so we don't need to explicitly create an attribute set in the first place.

The current implementation results in an attribute set like this

```
nix-repl> deprecations
{ gist-vim = { ... }; vim-jade = { ... }; vundle = { ... }; youcompleteme = { ... }; }

nix-repl> deprecations.gist-vim
{ old = «error: gist-vim was renamed to vim-gist. Please update to vim-gist.»; }
```

I guess the vim packages implementation evaluates all attributes of the plugins for some reason, which is why the `throw` works there. The `vim-plug` implementation never evaluates `gist-vim.old`. `gist-vim` is just an attrset, and it doesn't have the `rtp` attribute.

So in summary, to fix it you need to replace `deprecations` with

```nix
  deprecations = lib.mapAttrs (
    old: new: throw "${old} was renamed to ${new}. Please update to ${new}."
  )(
      builtins.fromJSON (builtins.readFile ./deprecated.json)
  )
```

**Comment by timokau** [comment]
Out of curiosity I also figured out why it works with vim packages. The implementation uses `lib.unique` on a list of plugins:
https://github.com/NixOS/nixpkgs/blob/4a3f9aced7ff35ff8cd7e021a542121a66396586/pkgs/misc/vim-plugins/vim-utils.nix#L318
To remove duplicates from the list, pairwise comparison is necessary. For comparison you need to evaluate the whole thing, which leads to evaluation of the `throw` expression.

**Review by jonringer — APPROVED** [review]
diff LGTM

let's users know what the new attr is
```
[1 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/83119
17 package built:
vimPlugins.Vundle-vim vimPlugins.YouCompleteMe vimPlugins.denite-nvim vimPlugins.deoplete-fish vimPlugins.deoplete-khard vimPlugins.deoplete-nvim vimPlugins.neoformat vimPlugins.nvim-lsp vimPlugins.vim-codefmt vimPlugins.vim-fugitive vimPlugins.vim-gist vimPlugins.vim-go vimPlugins.vim-lsc vimPlugins.vim-pug vimPlugins.vim-test vimPlugins.vimtex vimPlugins.yats-vim
```

I'll let @timokau click merge as he has more context on this PR

File: pkgs/misc/vim-plugins/update.py
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Its not *quite* what I had in mind. Now that we rewrite the file every time, there's not really a need to take the old file into consideration at all in `rewrite_input`. We could just re-generate an `plugin-names` file from scratch / from the data that was originally parsed. There would be no need to replace individual lines and worry about whitespace.

That's just a nitpick though, no reason to hold this off any longer.

**Review by timokau — APPROVED** [review]
Looks good to me! Thanks for sticking with this so long and putting up with all my requests. This is a bit improvement, much appreciated :+1:

**Comment by ryneeverett** [comment]
Thanks for all the great review @timokau!

> Really the only reliable way to make sure these PRs are really okay is to fully automate the updates and ideally have them run by a trusted entity.

Is there a roadmap for this? (Perhaps open a separate issue to track roadmap?) 

I'm wondering if we could use nixpkgs-update even though most of the packages probably aren't tracked by Repology. I would think it would be sufficient to just update the whole package set each time one of the tracked packages (for example, YouCompleteMe) is updated.

**Comment by jonringer** [comment]
> Is there a roadmap for this? (Perhaps open a separate issue to track roadmap?)

seeing as this is just a script that needs to be executed, we could just make a simple bot that runs the script on some interval

**Comment by timokau** [comment]
No roadmap, since nobody has been motivated enough to implement this yet I think. There are related issues though: https://github.com/ryantm/nixpkgs-update/issues/117, https://github.com/ryantm/nixpkgs-update/issues/12.

There are some exciting developments recently in nixpkgs-update: https://discourse.nixos.org/t/nixpkgs-update-automation-and-new-sources/5645

> seeing as this is just a script that needs to be executed, we could just make a simple bot that runs the script on some interval

It would be great to reuse the already trusted r-ryantm hardware, but that would also be a good alternative.

**Comment by jonringer** [comment]
nixpkgs-update IIRC only scrapes repology for version information, this script just directly scrapes from the repo's. So I don't think they have alignment on how they determine the need for a PR. 

@ryantm @bhipple correct me if I'm wrong. The question being: can nixpkgs-update also do vimplugins (I'm assuming it can, but it would be gathering the metadata off repology rather than github)

**Comment by ryantm** [comment]
I don't have an ideological opposition to nixpkgs-update doing things like running a script and making a PR that way. Though, it would be a very different branch of the code.

**Comment by ryantm** [comment]
If someone made a regular script that used hub and added it as a systemd service here https://github.com/nix-community/infra/blob/master/build01/nixpkgs-update.nix  We could consider including it under things that @r-ryantm does, but are not specifically using the nixpkgs-update Haskell code.

**Comment by jonringer** [comment]
> If someone made a regular script that used hub (to make the PR) and added it as a systemd service

I would be very okay with this, and in the PR description, just make a mention that it was generated using the script

**Pull request merged**
